core/sdk-core/src/main/java/software/amazon/awssdk/core/progress/listener/ProgressListener.java

```diff
@@ -0,0 +1,395 @@
+ * ProgressListener}s to be associated with that request. Then, throughout the lifecycle of the request,
+ * the Sdk
+ * will invoke the provided {@link ProgressListener}s when important events occur, like additional bytes being transferred,
+ * allowing you to monitor the ongoing progress of the transaction.
+ * <p>
+ * Each {@link ProgressListener} callback is invoked with an immutable {@link Context} object. Depending on the current
+ * lifecycle
+ * of the request, different {@link Context} objects have different attributes available (indicated by the provided context
+ * interface). Most notably, every callback is given access to the current {@link ProgressSnapshot}, which contains
+ * helpful progress-related methods like {@link ProgressSnapshot#transferredBytes()} and {@link
+ * ProgressSnapshot#ratioTransferred()}.
+ * <p>
+ * A successful transfer callback lifecycle is sequenced as follows:
+ * <ol>
+ *     <li>{@link #requestPrepared(Context.RequestPrepared)} - A new Request has been initiated. This method is called
+ *     exactly once per transfer.</li>
+ *     <ul>Available context attributes:
+ *         <li>{@link Context.RequestPrepared#request()}</li>
+ *         <li>{@link Context.RequestPrepared#progressSnapshot()}</li>
+ *     </ul>
+ *     <li>{@link #requestBytesSent(Context.RequestBytesSent)} - Additional bytes have been sent. This
+ *     method may be called many times per request, depending on the request payload size and I/O buffer sizes.
+ *     <li>{@link #responseBytesReceived(Context.ResponseBytesReceived)} - Additional bytes have been received. This
+ *     method may be called many times per request, depending on the response payload size and I/O buffer sizes.
+ *     <li>{@link #executionSuccess(Context.ExecutionSuccess)} - The transfer has completed successfully. This method is called
+ *     exactly once for a successful transfer.</li>
+ * </ol>
+ * For every failed attempt {@link #attemptFailure(Context.AttemptFailure)}  will be called exactly once.
+ *
+ * <p>
+ * There are a few important rules and best practices that govern the usage of {@link ProgressListener}s:
+ * <ol>
+ *     <li>{@link ProgressListener} implementations should not block, sleep, or otherwise delay the calling thread. If you need
+ *     to perform blocking operations, you should schedule them in a separate thread or executor that you control.</li>
+ *     <li>Be mindful that {@link #requestBytesSent(Context.RequestBytesSent)} or
+ *     {@link #responseBytesReceived(Context.ResponseBytesReceived)}
+ *     may be called extremely often for large payloads
+ *     (subject to I/O buffer sizes). Be careful in implementing expensive operations as a side effect. Consider rate-limiting
+ *     your side effect operations, if needed.</li>
+ *     <li>{@link ProgressListener}s may be invoked by different threads. If your {@link ProgressListener} is stateful,
+ *     ensure that it is also thread-safe.</li>
```

Comment:
This implies we don't have a happens-before relationship for each of the methods. Is that true? If so, for which methods? Is it true for HTTP/1 as well as HTTP/2?

Reply:
Yeah do we need happens-before links to progress methods though ? These methods are ideally triggered only when the previous event has completed and there are further progress updates.

Reply:
I thought of it being this way for all methods in all protocols. Let me know if we need to re-think that